File: models/trader.py
import random
import time
import threading
from datetime import datetime
import numpy as np

from models.order import Order, OrderSide

class Trader:
    """
    Simulated trading bot that generates orders
    """
    
    def __init__(self, trader_id, initial_cash, symbols, engine):
        """
        Initialize a trading bot
        
        Args:
            trader_id (str): Unique identifier for the trader
            initial_cash (float): Starting cash amount
            symbols (list): List of symbols to trade
            engine: Reference to the trading engine
        """
        self.trader_id = trader_id
        self.initial_cash = initial_cash
        self.cash = initial_cash
        self.symbols = symbols
        self.engine = engine
        
        # Portfolio tracking
        self.positions = {symbol: 0 for symbol in symbols}  # Share positions
        self.average_costs = {symbol: 0.0 for symbol in symbols}  # Average cost basis
        
        # Trading statistics
        self.orders_sent = 0
        self.orders_filled = 0
        self.total_volume = 0
        
        # Trading parameters
        self.min_order_size = 10
        self.max_order_size = 100
        self.price_volatility = 0.02  # 2% price variation
        self.order_frequency = 0.5  # Seconds between orders
        self.market_price_cache = {symbol: 100.0 for symbol in symbols}  # Starting prices
        
        # Threading
        self.is_active = False
        self.trade_thread = None
    
    def start_trading(self):
        """Start the trading bot in a separate thread"""
        if not self.is_active:
            self.is_active = True
            self.trade_thread = threading.Thread(target=self._trading_loop, daemon=True)
            self.trade_thread.start()
    
    def stop_trading(self):
        """Stop the trading bot"""
        self.is_active = False
        if self.trade_thread and self.trade_thread.is_alive():
            self.trade_thread.join(timeout=1.0)
    
    def _trading_loop(self):
        """Main trading loop that runs in a separate thread"""
        while self.is_active:
            try:
                # Random delay between orders
                delay = np.random.exponential(self.order_frequency)
                time.sleep(delay)
                
                if not self.is_active:
                    break
                
                # Generate a random order
                self._generate_order()
                
            except Exception as e:
                print(f"Error in trading loop for {self.trader_id}: {e}")
                time.sleep(1)  # Brief pause on error
    
    def _generate_order(self):
        """Generate and submit a random order"""
        if not self.symbols:
            return
        
        # Choose random symbol
        symbol = random.choice(self.symbols)
        
        # Get current market price estimate
        market_price = self._estimate_market_price(symbol)
        
        # Decide order side (buy/sell) with some bias based on position
        side = self._decide_order_side(symbol)
        
        # Generate order parameters
        quantity = random.randint(self.min_order_size, self.max_order_size)
        
        # Generate price with some randomness around market price
        price_variation = random.gauss(0, self.price_volatility)
        
        if side == OrderSide.BUY:
            # Buyers typically bid below market price
            price = market_price * (1 - abs(price_variation))
        else:
            # Sellers typically ask above market price
            price = market_price * (1 + abs(price_variation))
        
        # Round price to 2 decimal places
        price = round(price, 2)
        
        # Check if we can afford the order (for buy orders)
        if side == OrderSide.BUY and quantity * price > self.cash:
            # Reduce quantity or skip order
            affordable_quantity = int(self.cash / price)
            if affordable_quantity < self.min_order_size:
                return  # Can't afford minimum order
            quantity = affordable_quantity
        
        # Check if we have enough shares to sell
        if side == OrderSide.SELL and quantity > self.positions[symbol]:
            if self.positions[symbol] < self.min_order_size:
                return  # Not enough shares to sell
            quantity = self.positions[symbol]
        
        # Create and submit order
        order = Order(self.trader_id, symbol, side, quantity, price)
        self.engine.submit_order(order)
        self.orders_sent += 1
    
    def _estimate_market_price(self, symbol):
        """Estimate current market price based on recent trades and order book"""
        # Get recent trades for price discovery
        recent_trades = self.engine.get_recent_trades_for_symbol(symbol, 5)
        
        if recent_trades:
            # Use volume-weighted average of recent trades
            total_value = sum(trade['price'] * trade['quantity'] for trade in recent_trades)
            total_volume = sum(trade['quantity'] for trade in recent_trades)
            if total_volume > 0:
                self.market_price_cache[symbol] = total_value / total_volume
        else:
            # If no recent trades, use order book mid-price or random walk
            orderbook = self.engine.get_orderbook(symbol)
            best_bid = orderbook.get_best_bid()
            best_ask = orderbook.get_best_ask()
            
            if best_bid and best_ask:
                self.market_price_cache[symbol] = (best_bid.price + best_ask.price) / 2
            else:
                # Random walk from current price
                change = random.gauss(0, 0.01)  # 1% daily volatility
                self.market_price_cache[symbol] *= (1 + change)
                self.market_price_cache[symbol] = max(1.0, self.market_price_cache[symbol])
        
        return self.market_price_cache[symbol]
    
    def _decide_order_side(self, symbol):
        """Decide whether to buy or sell based on current position and market conditions"""
        position = self.positions[symbol]
        
        # If we have a large position, bias toward selling
        if position > 500:
            return OrderSide.SELL if random.random() < 0.7 else OrderSide.BUY
        # If we have no position, bias toward buying
        elif position == 0:
            return OrderSide.BUY if random.random() < 0.7 else OrderSide.SELL
        # Otherwise, random
        else:
            return random.choice([OrderSide.BUY, OrderSide.SELL])
    
    def on_order_filled(self, order, fill_quantity, fill_price):
        """
        Callback when an order is filled (called by the engine)
        
        Args:
            order: The filled order
            fill_quantity: Quantity filled
            fill_price: Price of the fill
        """
        symbol = order.symbol
        
        if order.side == OrderSide.BUY:
            # Update cash and position
            cost = fill_quantity * fill_price
            self.cash -= cost
            
            # Update average cost basis
            old_position = self.positions[symbol]
            old_cost_basis = self.average_costs[symbol] * old_position
            new_cost_basis = old_cost_basis + cost
            new_position = old_position + fill_quantity
            
            self.positions[symbol] = new_position
            if new_position > 0:
                self.average_costs[symbol] = new_cost_basis / new_position
            
        else:  # SELL
            # Update cash and position
            proceeds = fill_quantity * fill_price
            self.cash += proceeds
            self.positions[symbol] -= fill_quantity
            
            # If position goes to zero, reset average cost
            if self.positions[symbol] == 0:
                self.average_costs[symbol] = 0.0
        
        self.orders_filled += 1
        self.total_volume += fill_quantity
    
    def get_portfolio_value(self):
        """Calculate current portfolio value based on market prices"""
        portfolio_value = self.cash
        
        for symbol, position in self.positions.items():
            if position > 0:
                market_price = self._estimate_market_price(symbol)
                portfolio_value += position * market_price
        
        return portfolio_value
    
    def get_total_pnl(self):
        """Calculate total profit/loss"""
        return self.get_portfolio_value() - self.initial_cash
    
    def get_position_pnl(self, symbol):
        """Calculate P&L for a specific position"""
        position = self.positions[symbol]
        if position == 0:
            return 0.0
        
        market_price = self._estimate_market_price(symbol)
        market_value = position * market_price
        cost_basis = position * self.average_costs[symbol]
        
        return market_value - cost_basis
    
    def get_trading_stats(self):
        """Get trading statistics"""
        return {
            'trader_id': self.trader_id,
            'cash': self.cash,
            'portfolio_value': self.get_portfolio_value(),
            'total_pnl': self.get_total_pnl(),
            'orders_sent': self.orders_sent,
            'orders_filled': self.orders_filled,
            'total_volume': self.total_volume,
            'positions': self.positions.copy(),
            'fill_rate': self.orders_filled / max(1, self.orders_sent)
        }

File: cpp_hft/src/trader.cpp
#include "trader.h"
#include "engine.h"
#include <QMutexLocker>
#include <QDebug>
#include <QVariant>
#include <cmath>

Trader::Trader(const QString& traderId, double initialCash, 
               const QStringList& symbols, TradingEngine* engine, QObject* parent)
    : QObject(parent)
    , traderId(traderId)
    , initialCash(initialCash)
    , cash(initialCash)
    , symbols(symbols)
    , engine(engine)
    , ordersSent(0)
    , ordersFilled(0)
    , totalVolume(0)
    , minOrderSize(10)
    , maxOrderSize(100)
    , priceVolatility(0.02)  // 2% price variation
    , orderFrequencyMs(50)   // 50ms between orders for HFT speed
    , active(false)
    , tradingTimer(new QTimer(this))
    , random(QRandomGenerator::global())
{
    // Initialize positions and market price cache
    for (const QString& symbol : symbols) {
        positions[symbol] = 0;
        averageCosts[symbol] = 0.0;
        marketPriceCache[symbol] = 100.0; // Starting price
    }
    
    // Connect timer
    connect(tradingTimer, &QTimer::timeout, this, &Trader::generateOrder);
    
    // Set random interval for trading
    tradingTimer->setSingleShot(false);
}

Trader::~Trader() {
    stopTrading();
}

QMap<QString, int> Trader::getPositions() const {
    QMutexLocker locker(&mutex);
    return positions;
}

void Trader::startTrading() {
    QMutexLocker locker(&mutex);
    
    if (!active) {
        active = true;
        // Start with random delay
        int delay = random->bounded(100, orderFrequencyMs);
        tradingTimer->start(delay);
    }
}

void Trader::stopTrading() {
    QMutexLocker locker(&mutex);
    
    if (active) {
        active = false;
        tradingTimer->stop();
    }
}

double Trader::getPortfolioValue() const {
    QMutexLocker locker(&mutex);
    
    double portfolioValue = cash;
    
    for (auto it = positions.begin(); it != positions.end(); ++it) {
        const QString& symbol = it.key();
        int position = it.value();
        
        if (position > 0) {
            double marketPrice = estimateMarketPrice(symbol);
            portfolioValue += position * marketPrice;
        }
    }
    
    return portfolioValue;
}

double Trader::getTotalPnL() const {
    return getPortfolioValue() - initialCash;
}

double Trader::getPositionPnL(const QString& symbol) const {
    QMutexLocker locker(&mutex);
    
    int position = positions.value(symbol, 0);
    if (position == 0) {
        return 0.0;
    }
    
    double marketPrice = estimateMarketPrice(symbol);
    double marketValue = position * marketPrice;
    double costBasis = position * averageCosts.value(symbol, 0.0);
    
    return marketValue - costBasis;
}

QMap<QString, QVariant> Trader::getTradingStats() const {
    QMutexLocker locker(&mutex);
    
    QMap<QString, QVariant> stats;
    stats["trader_id"] = traderId;
    stats["cash"] = cash;
    stats["portfolio_value"] = getPortfolioValue();
    stats["total_pnl"] = getTotalPnL();
    stats["orders_sent"] = ordersSent;
    stats["orders_filled"] = ordersFilled;
    stats["total_volume"] = totalVolume;
    stats["fill_rate"] = ordersFilled / qMax(1, ordersSent);
    
    return stats;
}

void Trader::onOrderFilled(OrderPtr order, int fillQuantity, double fillPrice) {
    QMutexLocker locker(&mutex);
    
    updatePosition(order->getSymbol(), order->getSide(), fillQuantity, fillPrice);
    ordersFilled++;
    totalVolume += fillQuantity;
}

void Trader::generateOrder() {
    if (!active || symbols.isEmpty()) {
        return;
    }
    
    QMutexLocker locker(&mutex);
    
    // Choose random symbol
    QString symbol = symbols[random->bounded(symbols.size())];
    
    // Get current market price estimate
    double marketPrice = estimateMarketPrice(symbol);
    
    // Decide order side (buy/sell) with some bias based on position
    OrderSide side = decideOrderSide(symbol);
    
    // Generate order parameters
    int quantity = random->bounded(minOrderSize, maxOrderSize + 1);
    
    // Generate price with some randomness around market price
    double priceVariation = random->generateDouble() * priceVolatility * 2 - priceVolatility;
    double price;
    
    if (side == OrderSide::BUY) {
        // Buyers typically bid below market price
        price = marketPrice * (1 - std::abs(priceVariation));
    } else {
        // Sellers typically ask above market price
        price = marketPrice * (1 + std::abs(priceVariation));
    }
    
    // Round price to 2 decimal places
    price = std::round(price * 100.0) / 100.0;
    
    // Check if we can afford the order (for buy orders)
    if (side == OrderSide::BUY && quantity * price > cash) {
        int affordableQuantity = static_cast<int>(cash / price);
        if (affordableQuantity < minOrderSize) {
            // Schedule next order
            tradingTimer->start(random->bounded(100, orderFrequencyMs));
            return;
        }
        quantity = affordableQuantity;
    }
    
    // Check if we have enough shares to sell
    if (side == OrderSide::SELL && quantity > positions.value(symbol, 0)) {
        int availableShares = positions.value(symbol, 0);
        if (availableShares < minOrderSize) {
            // Schedule next order
            tradingTimer->start(random->bounded(100, orderFrequencyMs));
            return;
        }
        quantity = availableShares;
    }
    
    // Create and submit order
    auto order = std::make_shared<Order>(traderId, symbol, side, quantity, price);
    
    locker.unlock(); // Unlock before calling engine
    
    engine->submitOrder(order);
    
    locker.relock();
    ordersSent++;
    
    // Schedule next order with random delay
    int nextDelay = random->bounded(100, orderFrequencyMs);
    tradingTimer->start(nextDelay);
}

double Trader::estimateMarketPrice(const QString& symbol) const {
    // This method would normally query the engine for recent trades and order book
    // For simplicity, we'll use a basic random walk model
    
    double currentPrice = marketPriceCache.value(symbol, 100.0);
    
    // Random walk with small steps
    double change = (random->generateDouble() - 0.5) * 0.02; // ±1% change
    currentPrice *= (1 + change);
    currentPrice = qMax(1.0, currentPrice); // Minimum price of $1
    
    // Update cache (note: this is not thread-safe, but acceptable for simulation)
    const_cast<Trader*>(this)->marketPriceCache[symbol] = currentPrice;
    
    return currentPrice;
}

OrderSide Trader::decideOrderSide(const QString& symbol) {
    int position = positions.value(symbol, 0);
    
    // If we have a large position, bias toward selling
    if (position > 500) {
        return (random->generateDouble() < 0.7) ? OrderSide::SELL : OrderSide::BUY;
    }
    // If we have no position, bias toward buying
    else if (position == 0) {
        return (random->generateDouble() < 0.7) ? OrderSide::BUY : OrderSide::SELL;
    }
    // Otherwise, random
    else {
        return (random->generateDouble() < 0.5) ? OrderSide::BUY : OrderSide::SELL;
    }
}

void Trader::updatePosition(const QString& symbol, OrderSide side, int quantity, double price) {
    if (side == OrderSide::BUY) {
        // Update cash and position
        double cost = quantity * price;
        cash -= cost;
        
        // Update average cost basis
        int oldPosition = positions.value(symbol, 0);
        double oldCostBasis = averageCosts.value(symbol, 0.0) * oldPosition;
        double newCostBasis = oldCostBasis + cost;
        int newPosition = oldPosition + quantity;
        
        positions[symbol] = newPosition;
        if (newPosition > 0) {
            averageCosts[symbol] = newCostBasis / newPosition;
        }
    } else { // SELL
        // Update cash and position
        double proceeds = quantity * price;
        cash += proceeds;
        positions[symbol] -= quantity;
        
        // If position goes to zero, reset average cost
        if (positions[symbol] == 0) {
            averageCosts[symbol] = 0.0;
        }
    }
}